hail/python/hailtop/aiotools/stream.py

```diff
@@ -125,3 +129,73 @@ def blocking_readable_stream_to_async(thread_pool: ThreadPoolExecutor, f: Binary
 
 def blocking_writable_stream_to_async(thread_pool: ThreadPoolExecutor, f: BinaryIO) -> _WritableStreamFromBlocking:
     return _WritableStreamFromBlocking(thread_pool, f)
+
+
+class BlockingQueueReadableStream(io.RawIOBase):
+    # self.closed and self.close() must be multithread safe, because
+    # they can be accessed by both the stream reader and writer which
+    # are in different threads.
+    def __init__(self, q: janus.Queue):
+        super().__init__()
+        self._q = q
+        self._saw_eos = False
+        self._closed = False
+        self._unread = b''
+
+    def readable(self) -> bool:
+        return True
+
+    def readinto(self, b: bytearray) -> int:
+        if self._closed:
+            raise ValueError('read on closed stream')
+        if self._saw_eos:
+            return 0
+
+        while not self._unread:
```

Comment:
Isn't this equivalent to an `if`?

Reply:
No.  sync_q can return `b''`, and empty bytes-like object.

However, I changed this code anyway.  Now write only puts non-empty objects in the q, so this can be an `if`.  I like that better, good suggestion.